File: pyppeteer/nbrender.py
# Copyright: Gergely Szerovay, 2019
# Licensed under the Apache License, version 2.0.

# 96 dpi

import asyncio
import pyppeteer
import sys
import os

def usage():
    print("USAGE:\npython nbrender.py ./notebook.html notebook.pdf\nOR\npython nbrender.py ./notebook.html notebook.png")

async def main():
    if len(sys.argv) != 3:
        usage()
        return
    
    htmlFile = sys.argv[1]
    outputFile = sys.argv[2]
    outputFileExt = outputFile[-4:]
    
    browser = await pyppeteer.launch()
    page = await browser.newPage()
    await page.goto('file://' + os.path.realpath(htmlFile))
    
    await page.setViewport({'width':1920, 'height': 100, 'deviceScaleFactor': 1})

    
    dimensions = await page.evaluate('''() => {
        return {
            deviceScaleFactor: window.devicePixelRatio,
            width: document.body.scrollWidth,
            height: document.body.scrollHeight
        }
    }''')

    pageHeightIn = (dimensions['height'] + 96 * 2) / 96
    print(f'Device scale factor: {dimensions["deviceScaleFactor"]}');
    print(f'Page width: {dimensions["width"]}px');
    print(f'Page height: {dimensions["height"]}px');
    
    if outputFileExt == '.png':
        await page.screenshot({'path': outputFile, 'fullPage': True})
    
    elif outputFileExt == '.pdf':          
        print(f'Page height: {pageHeightIn:.2f}in');
        if pageHeightIn > 200:
            print("\nWARNING: A page in a PDF document must be smaller than 200 inches.\n")

        await page.pdf({
            'path': outputFile,
            'printBackground': True,
            'margin': {'left': 0, 'top':0, 'right':0, 'bottom': 0},
            'width': 1920,
            'height': dimensions['height'] + 2 * 96
        })
    else:
        print('Invalid output file type, only .pdf and .png supported.')
    
    await browser.close()

asyncio.get_event_loop().run_until_complete(main())

File: phantomsj/nbrender.js
"use strict"

// Copyright: Gergely Szerovay, 2019
// Licensed under the Apache License, version 2.0.

var system = require('system'),
    page = require('webpage').create(),
    page2 = require('webpage').create(),
    htmlFile, outputFile, pageWidth2, pageHeightIn,
    dpi = 72,
    pageWidth = 1920,
    pageHeight = 100;

if (system.args.length < 3) {
    console.log('USAGE: phantomjs ./nbrender.js ./notebook.html notebook.pdf');
    phantom.exit(1);
} else {
    htmlFile = system.args[1];
    outputFile = system.args[2];

    page.paperSize = { width: pageWidth, height: pageHeight, margin: '0px' }
    page.viewportSize = { width: pageWidth, height: pageHeight };
    page.clipRect = { top: 0, left: 0, width: pageWidth, height: pageHeight };

    page.open(htmlFile, function (status) {
        // calculate the height of the rendered page at viewport width = 1920px
        if (status !== 'success') {
            console.log('Unable to load the HTML file!');
            phantom.exit(1);
        } else {
            window.setTimeout(function () {
                console.log('Page opened.');
                pageHeight = page.evaluate(function(){
                    return document.body.scrollHeight;
                });
                pageHeight += dpi * 2; // add some more space
                // verify page width
                pageWidth2 = page.evaluate(function(){
                    return document.body.scrollWidth;
                });
                console.log('Page width: ' + pageWidth2 + 'px');
                console.log('Page height: ' + pageHeight + 'px');
                pageHeightIn = Math.round(pageHeight/dpi)
                console.log('Page height: ' + pageHeightIn + 'in');

                if (pageHeightIn > 200 && outputFile.substr(-4) === ".pdf") {
                    console.log("\nWARNING: A page in a PDF document must be smaller than 200 inches.\n")
                }

                // reopen the page with the calculated page height
                page2.paperSize = { width: pageWidth, height: pageHeight, margin: '0px' }
                page2.viewportSize = { width: pageWidth, height: pageHeight };
                page2.clipRect = { top: 0, left: 0, width: pageWidth, height: pageHeight };

                console.log('Rendering page...');
                page2.open(htmlFile, function (status) {
                    window.setTimeout(function () {
                        page2.render(outputFile);
                        console.log('Done.');
                        phantom.exit();
                    }, 3000);
                });
            }, 3000);
        }
    });
}
